File: internal/apiclient/iac.go
package apiclient

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"

	"github.com/GLINCKER/levelrail/internal/iac"
)

// IaCFile is one resource file sent to plan or apply.
type IaCFile struct {
	Name    string `json:"name"`
	Content string `json:"content"`
}

// IaCRequest mirrors internal/api's iacRequest, the body of POST
// /api/v1/apply/plan and /api/v1/apply.
type IaCRequest struct {
	Files            []IaCFile         `json:"files"`
	Source           string            `json:"source,omitempty"`
	Project          string            `json:"project,omitempty"`
	Prune            bool              `json:"prune,omitempty"`
	Vars             map[string]string `json:"vars,omitempty"`
	Secrets          map[string]string `json:"secrets,omitempty"`
	NoDeploy         bool              `json:"no_deploy,omitempty"`
	ContinueOnError  bool              `json:"continue_on_error,omitempty"`
	ExpectedPlanHash string            `json:"expected_plan_hash,omitempty"`
}

// IaCIssuesError carries the line numbered validation issues of a 422.
type IaCIssuesError struct{ Issues []iac.Issue }

func (e *IaCIssuesError) Error() string {
	lines := make([]string, len(e.Issues))
	for i, is := range e.Issues {
		lines[i] = is.String()
	}
	return "invalid resource files:\n  " + strings.Join(lines, "\n  ")
}

func asIssues(err error) error {
	var api *APIError
	if !errors.As(err, &api) || api.StatusCode != http.StatusUnprocessableEntity {
		return err
	}
	var body struct {
		Issues []iac.Issue `json:"issues"`
	}
	if json.Unmarshal([]byte(api.Message), &body) != nil || len(body.Issues) == 0 {
		return err
	}
	return &IaCIssuesError{Issues: body.Issues}
}

// PlanIaC calls POST /api/v1/apply/plan.
func (c *Client) PlanIaC(ctx context.Context, req IaCRequest) (iac.Plan, error) {
	var out struct {
		Plan *iac.Plan `json:"plan"`
	}
	if err := c.do(ctx, http.MethodPost, "/api/v1/apply/plan", req, &out); err != nil {
		return iac.Plan{}, asIssues(err)
	}
	if out.Plan == nil {
		return iac.Plan{}, errors.New("plan response held no plan")
	}
	return *out.Plan, nil
}

// ApplyIaC calls POST /api/v1/apply.
func (c *Client) ApplyIaC(ctx context.Context, req IaCRequest) (iac.ApplyResult, error) {
	var out struct {
		Result *iac.ApplyResult `json:"result"`
	}
	if err := c.do(ctx, http.MethodPost, "/api/v1/apply", req, &out); err != nil {
		return iac.ApplyResult{}, asIssues(err)
	}
	if out.Result == nil {
		return iac.ApplyResult{}, errors.New("apply response held no result")
	}
	return *out.Result, nil
}

// ExportIaC calls GET /api/v1/export.
func (c *Client) ExportIaC(ctx context.Context, project, app string, includeEnvValues bool) (iac.ExportResult, error) {
	q := url.Values{}
	if project != "" {
		q.Set("project", project)
	}
	if app != "" {
		q.Set("app", app)
	}
	q.Set("include_env_values", fmt.Sprint(includeEnvValues))
	var out iac.ExportResult
	err := c.do(ctx, http.MethodGet, "/api/v1/export?"+q.Encode(), nil, &out)
	return out, err
}

File: web/src/queries/iac.ts
// Fetchers and hooks for the declarative resource endpoints
// (internal/api/iac.go): POST /api/v1/apply/plan, POST /api/v1/apply and
// GET /api/v1/export.

import { useMutation, useQueryClient } from '@tanstack/react-query'
import { appKeys } from './apps'
import { projectKeys } from './projects'
import { ApiError, readErrorMessage } from '../lib/apiError'

export type IacAction = 'create' | 'update' | 'delete' | 'noop' | 'error'
export type IacFieldOp = 'add' | 'change' | 'remove' | 'keep'

export interface IacFieldChange {
  path: string
  op: IacFieldOp
  old?: string
  new?: string
}

export interface IacChange {
  kind: string
  name: string
  scope?: string
  action: IacAction
  fields?: IacFieldChange[]
  kept?: IacFieldChange[]
  warnings?: string[]
  reason?: string
  denied?: boolean
  file?: string
  line?: number
}

export interface IacSummary {
  create: number
  update: number
  delete: number
  noop: number
  error: number
}

export interface IacPlan {
  source?: string
  prune?: boolean
  changes: IacChange[]
  summary: IacSummary
  hash: string
}

export interface IacIssue {
  file?: string
  path?: string
  line: number
  message: string
}

export type IacItemStatus = 'applied' | 'failed' | 'denied' | 'skipped' | 'noop'

export interface IacItemResult extends IacChange {
  status: IacItemStatus
  error?: string
}

export interface IacApplyResult {
  plan: IacPlan
  results: IacItemResult[]
  applied: number
  failed: number
  skipped: number
}

export interface IacFile {
  name: string
  content: string
}

export interface IacRequest {
  files: IacFile[]
  source?: string
  project?: string
  prune?: boolean
  no_deploy?: boolean
  continue_on_error?: boolean
  expected_plan_hash?: string
}

export interface IacExportFile {
  name: string
  kind: string
  content: string
}

export interface IacExportResult {
  files: IacExportFile[]
  warnings?: string[]
}

// Thrown for a 422: the files did not validate, with file and line per issue.
export class IacIssuesError extends ApiError {
  readonly issues: IacIssue[]

  constructor(issues: IacIssue[]) {
    super(422, 'The resource files are invalid.')
    this.name = 'IacIssuesError'
    this.issues = issues
  }
}

async function iacRequest<T>(
  url: string,
  init: RequestInit | undefined,
  what: string,
): Promise<T> {
  const res = await fetch(url, init)
  if (res.status === 422) {
    const body = (await res.json().catch(() => null)) as {
      issues?: IacIssue[]
    } | null
    if (body?.issues?.length) throw new IacIssuesError(body.issues)
  }
  if (!res.ok) {
    throw new ApiError(
      res.status,
      await readErrorMessage(res, `${what} failed: ${res.status}`),
    )
  }
  return (await res.json()) as T
}

function postJson(body: IacRequest): RequestInit {
  return {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  }
}

export function usePlanIac() {
  return useMutation<IacPlan, ApiError, IacRequest>({
    mutationFn: async (req) => {
      const out = await iacRequest<{ plan: IacPlan }>(
        '/api/v1/apply/plan',
        postJson(req),
        'plan',
      )
      return out.plan
    },
  })
}

export function useApplyIac() {
  const queryClient = useQueryClient()
  return useMutation<IacApplyResult, ApiError, IacRequest>({
    mutationFn: async (req) => {
      const out = await iacRequest<{ result: IacApplyResult }>(
        '/api/v1/apply',
        postJson(req),
        'apply',
      )
      return out.result
    },
    onSettled: () => {
      void queryClient.invalidateQueries({ queryKey: appKeys.all })
      void queryClient.invalidateQueries({ queryKey: projectKeys.all })
    },
  })
}

export interface IacExportParams {
  project?: string
  app?: string
  includeEnvValues: boolean
}

export function useExportIac() {
  return useMutation<IacExportResult, ApiError, IacExportParams>({
    mutationFn: (params) => {
      const q = new URLSearchParams()
      if (params.project) q.set('project', params.project)
      if (params.app) q.set('app', params.app)
      q.set('include_env_values', String(params.includeEnvValues))
      return iacRequest<IacExportResult>(
        `/api/v1/export?${q.toString()}`,
        undefined,
        'export',
      )
    },
  })
}
